R/pkg-knitr.R

```diff
@@ -0,0 +1,170 @@
+#' knit print polars DataFrame
+#'
+#' Mimics python-polars' NotebookFormatter
+#' for HTML outputs.
+#'
+#' Outputs HTML tables if the output format is HTML
+#' and the document's `df_print` option is not `"default"` or `"tibble"`.
+#'
+#' Or, the output format can be enforced with R's `options` function as follows:
+#'
+#' - `options(polars.df_print = "default")` for the default print method.
+#' - `options(polars.df_print = "html")` for the HTML table.
+#' @name knit_print.DataFrame
+#' @param x a polars DataFrame to knit_print
+#' @param ... additional arguments, not used
+#' @keywords DataFrame
+#' @export
+knit_print.DataFrame <- function(x, ...) {
+  .print_opt <- getOption("polars.df_print", "auto")
+  .rmd_df_print <- knitr::opts_knit$get("rmarkdown.df_print")
+  if (.print_opt == "html" ||
+    (.print_opt != "default" &&
+      !isTRUE(.rmd_df_print %in% c("default", "tibble")) &&
+      knitr::is_html_output())) {
+    x |>
+      to_html_table() |>
+      knitr::raw_html() |>
+      knitr::asis_output()
+  } else {
+    print(x)
+  }
+}
+
+#' Generate HTML table from a DataFrame
+#' @param x DataFrame
+#' @param max_cols an integer of maximum number of columns to display
+#' @param max_rows an integer of maximum number of rows to display
+#' @return A string of HTML code
+#' @examples
+#' to_html_table(mtcars, 3, 3)
+#' @export
```

Comment:
Do we want to export this function? It's great to have internally, but do we want to commit to maintaining a table-making functions that users may end up using long term?

Reply:
Also, out of curiosity, what made you decide against `knitr::kable(format="html")`? I guess the output wouldn't match pypolars exactly, but there would be no maintenance and some "guarantees" that it plays well with Rmarkdown and Quarto, since so many users depend on it and it's well tested upstream.

To be clear, I'm not advocating; just curious.

Reply:
> Do we want to export this function? It's great to have internally, but do we want to commit to maintaining a table-making functions that users may end up using long term?

Good point. We probably don't need to export this.
(Users who want to use it can use it even if it is not public)

> what made you decide against knitr::kable(format="html")?

Short answer, simply because I like python-polars HTML tables.

I found the following document I recently created with Quarto to be a very good python polars table display.
https://eitsupi.github.io/querying-with-prql/indexing#filtering-rows

- Show types
- Show shape
- Automatically collapses the middle and displays the edges of the table

BTW, the ability to detect the `df_print` option and switch to output using `kable` can be implemented in the future.

Reply:
yeah, that document looks awesome. Let me know if/when you'd need me to install and try your branch.

Reply:
Thanks.
You do not need you to check the output locally as we can check output from the snapshot tests.